### PR TITLE
Helpers 2.1

### PR DESCRIPTION
**Pull request by alexAubin**
## The problem

Helperz bad

## Solution

Add a new "2.1" version for the helpers:

~25ish % of the code removed compared to v1

#### Important changes, renames etc

- [x] **General remarks regarding helper argument management**
   - Support for so-called "legacy" positional arg was removed in getopts to try to enforce full keyword argument usage (`--foo=bar`) and have as much of a uniform syntax across all helpers as possible
   - However, when refactoring, many simplifications were made leading to a lot of helpers only requiring a single arg, which is now handled as a positional
   - The result is that **many helpers are much simpler**, usually requiring **no args, or a single arg via a positional** (for example: `ynh_safe_rm /path/to/file`)
   - Keyword arguments become sort of "the exception", for complex helpers requiring many important args (such as `ynh_setup_source`, `ynh_add_config`, `ynh_systemctl`, etc...)
 - [x] **General remarks on helper naming conventions**
   - aiming to uniformize naming convention, we converged to basically having a `ynh_{tech}` "namespacing" :
   - "tech" versions are defined via global `${tech}_version` variables (such as `$php_version`, `$nodejs_version`, `$ruby_version`, `$composer_version`, ...)
   - helpers to install/remove "technologies" are named `ynh_{tech}_install` / `ynh_{tech}_remove` (with tech = nodejs, ruby, go, composer, apt, ...)
   - similar stuff for database helpers : `ynh_{dbtype}_shell`, `ynh_{dbtype}_dump_db` (with dbtype = mysql, psql, ... mongodb? redis?)
   - config-related helpers are named `ynh_config_add/remove_{tech}` (with tech = nginx, phpfpm, logrotate, fail2ban, ...), for example `ynh_config_add_nginx`
   - try to prefix all the internal / low-level helpers with `_` (i.e. `_ynh_foo_bar`) ?
- [x] Backup/restore
    - `ynh_restore` is renamed to `ynh_restore_everything`. 
    - `ynh_restore_file` is renamed to `ynh_restore` to be symmetric with `ynh_backup`. 
    - Single positional argument, no more `--dest_dir`, `--not_mandatory` or `--is_big`.
    - The helper handles the logic of checking if the path is `$data_dir` or `/var/log/$app`, in which case they are ignore during safety-backup-before-upgrade or if `do_not_backup_data=1`  (standardizing the old, removed "is_big" arg)
    - The core will now automatically delete `/var/log/$app` during `yunohost app remove` if `--purge` is used. The app remove script should not remove `/var/log/$app/` (otherwise this means it gets deleted during removal of the app prior to restoring the safety-backup-before-upgrade ...)
- [x] Fail2ban
    - remove unecessary/unused `--max_retry` and `--ports` options
    - remove `--use_template`: just generate the template "on-the-fly" if `--failregex`/`--logpath` are provided, or use the f2b_stuff template files from the `conf/` directory otherwise 
- [x] APT
    - simplify bloated, unecessarily complex code
    - rename helpers to `ynh_apt_install_dependencies`, `ynh_apt_remove_dependencies`,  `ynh_apt_install_dependencies_from_extra_repository`.
    - misc internal helpers such as `_ynh_apt_package_version` are prefixed with `_`
- [x] mysql/psql:
    - no keyword arg, full positional arg
    - remove `ynh_foo_setup_db` and `ynh_foo_remove_db` (the other helpers are enough)
    - replace `ynh_foo_connect_as`/`ynh_foo_execute*` with a single `ynh_foosql_db_shell` that reads stdin 
    - use `$db_user`, `$db_pwd`, `$db_name` as default values for user, password and database args
- [x] `ynh_setup_source`
    - `--full_replace` now to be a boolean, no need to write `--full_replace=1`
    - drop support for underused `sources/extra_files/$sourceid` mechanism, only 7ish apps actually do this and they manually cp their stuff instead of relying on the helper's behavior
    - move the `sources/patches/` mechanism to just `patches/` (at the root of the app dir)
- [x] PHP / composer
    - Rename `$phpversion` to `$php_version` for consistency with `$nodejs_version`, `$ruby_version` etc
    - Remove `--phpversion` and `--composerversion` (helpers will use the global `$php_version` / `$composer_version` to be defined prior to calling those helpers)
    - Remove `--usage`/`--footprint` (can still be customized by defining `fpm_usage`/`footprint` setting, otherwise defaults to "low" which the vast majority of apps do. It is encourage to just trash manually defining these settings apart if you really have reasons to define the usage/footprint as medium/high by default...)
    - Remove every arg in `ynh_composer_install` and `ynh_composer_exec` (apart for the composer command to run for `ynh_composer_exec`, using positional). The old `--workdir` is now automatically `$install_dir` unless `$composer_workdir` is defined prior to calling this helper.
    - `ynh_composer_install` now ONLY `wget`s the actual `composer.phar` directly instead of calling the unnecessarily complex official composer install script. `ynh_composer_exec` needs to be explicitly called for the actual "install" of the vendor/dependencies
    - `ynh_composer_exec` *runs as `$app` instead of root*, unless `composer_user=root` is defined prior to calling this helper
- [x] Rework `ynh_install_nodejs`/`ruby`/`go`
    - no more --version arg, just define `$nodejs_version`, `$ruby_version`, `$go_version` as global prior to calling it
    - Tweak verbosity such that packagers are not tempted to slap an `ynh_exec_warn_less` in front of it ...
    - Remove the need to call `ynh_use_nodjs`/`ruby`/`go`, it's just called at the end of the `ynh_{tech}_install` helper
    - Export the modified `PATH` variable such that one can call `npm`, `node`, `ruby`, `go` directly ... instead of the clumsy `$ynh_npm`/`node` var/aliases.
    - Because the `PATH` is exported and `ynh_exec_as_app` now uses `sudo -E`, there's no need to pass `$ynh_node_load_PATH` when calling npm/node/ruby/go/...
    - Rework `ynh_node_load_PATH` : replaced by `PATH_with_nodejs` such that we write `Environment="PATH=__PATH_WITH_NODEJS__"` in the systemd config
- [x] Rework `ynh_use_logrotate` / `ynh_remove_logrotate`
    - renamed to `ynh_config_add/remove_logrotate` for consistency with other helpers
    - Single position argument (the logfile(s)). No `--specific_user` option anymore (if logrotate complains, then the log dir permissions should be fixed)
- [x] Misc helper renaming/simplification
    - the remaining logging helpers (`ynh_print_info`, `ynh_print_warn`, `ynh_script_progression`) now use a positional arg (no need for `--message=` anymore). Drop the weight mechanism for `ynh_script_progression` which doesn't make much sense since packaging v2 anyway
    - `ynh_exec_as $app`  becomes `ynh_exec_as_app`, (with `-E` option in sudo, such that exported env variables such as `PATH` are propagated)
    - `ynh_secure_remove --file=/foo/bar` becomes `ynh_safe_rm /foo/bar` (single positional arg)
    - `ynh_exec_warn_less` becomes `ynh_hide_warnings`
    - `ynh_replace_string` becomes `ynh_replace --match=foo --replace=bar --file=/some/path`
    - `ynh_systemd_action`: 
        - renamed to `ynh_systemctl`
        - `--service_name` arg renamed to `--service` to be consistent with other commands
        - `--line_match` arg renamed to `--wait_until`
    - `ynh_compare_current_package_version --comparison lt --version X.Y~ynhZ` becomes 
`ynh_app_upgrading_from_version_before` and `ynh_app_upgrading_from_version_before_or_equal_to`
    - Replace the clumsy `if [ "$upgrade_type" == "UPGRADE_APP" ]` with a simpler `if ynh_app_upstream_version_changed` ...
    - Replace `if [ "${PACKAGE_CHECK_EXEC:-0}" -eq 1 ]` with simpler `if ynh_in_ci_tests`
    - Use positional args for `ynh_normalize_url_path`

#### Legacy cleanups

- [x] backport most of the bookworm changes (remove unused stuff)
- [x] remove unecessary `--app=$app` in internals ynh_app_setting_get/set calls
- [x] remove support for `__NAME__` and `__NAMETOCHANGE__` replaced by $app in templates
- [x] Drop support for old .src format in ynh_setup_source 
- [x] Remove legacy `ynh_legacy_permissions_exists` and `ynh_legacy_permissions_delete_all`
- [x] Remove legacy `ynh_webpath_available` and `ynh_webpath_register`
- [x] Remove legacy `ynh_psql_test_if_first_run`
- [x] Remove legacy `--nonappend`/`--non-append` arg in logrotate helper
- [x] Remove legacy `ynh_backup_before_upgrade` and `ynh_restore_upgradebackup`, handled by core in packaging v2
- [x] Remove legacy `ynh_find_port`, `ynh_port_available`, `ynh_validate_ip4/6`, keep only `ynh_validate_ip`
- [x] Remove `ynh_get_debian_release`, apps should just use `$YNH_DEBIAN_VERSION`
- [x] Remove `ynh_require_ram` and unused `--ignore_swap` and `--only_swap` options in `ynh_get_ram`
- [x] Simplify `ynh_app_upstream_version` and `ynh_read_manifest`
- [x] Remove legacy/unecessary/underused helpers: `ynh_print_log`, `ynh_print_err`, `ynh_exec_err`, `ynh_exec_quiet`, `ynh_exec_fully_quiet`, `ynh_print_OFF`, `ynh_print_ON`
- [x] Remove a whole bunch of unused args in mongo helpers..
- [x] Remove unused options `--label`/`--show_tile`/`--protected` in `ynh_permission_update`
- [x] Remove old internal `ynh_render_template`, should use `ynh_add_config --jinja` instead

## PR Status

...

## How to test

...

### CODE REVIEW

**Comment by Thovi98** [comment]
>  Rename ynh_secure_remove --file to ynh_safe_rm --target

For my knowledge, is there a specific reason for this? :)

**Comment by alexAubin** [comment]
>For my knowledge, is there a specific reason for this? :)

Not really, mainly personal taste I suppose, I just found it super cringy that it's called "secure" as it's about "security" (as against an external thread) rather than "safety" (ie like backup) and the fact that it's "remove" instead of "rm" which may somewhat be confusing like "is it different than rm then?"

**Comment by ericgaspar** [comment]
Amazing cleanup job! well done

**Comment by orhtej2** [comment]
Will removal of `app.src` support break restoring backups of v1-packaged apps?

Other than that amazing stuff, do we plan to automagically retrofit syntax changes to the whole catalogue?

File: helpers/helpers.v2.1.d/backup [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+# upgrade fails and the script is removed prior to restoring the backup, the
+# data/logs are not destroyed.
+#
+ynh_backup() {
+
+    local target="$1"
+    local is_data=false
+
+    # If the path starts with /var/log/$app or $data_dir
+    if ([[ -n "${app:-}" ]] && [[ "$target" == "/var/log/$app*" ]]) || ([[ -n "${data_dir:-}" ]] && [[ "$target" == "$data_dir*" ]])
```
Comment (Salamandar) [inline helpers/helpers.v2.1.d/backup:30]: > --is_big behavior is replaced by checking if the path is $data_dir (or a child) or /var/log/$app

Yeah that might be a bit restrictive and what if `$data_dir/some_important_config_file.json` ?

Comment (Salamandar) [reply helpers/helpers.v2.1.d/backup:30]: Also packagingv3™ but those directories could be declared in manifest.

File: helpers/helpers.v2.1.d/utils [hunk 2]
```diff
@@ -126,117 +126,78 @@ fi
...
-#
 # The helper will:
 # - Download the specific URL if there is no local archive
 # - Check the integrity with the specific sha256 sum
 # - Uncompress the archive to `$dest_dir`.
 #   - If `in_subdir` is true, the first level directory of the archive will be removed.
 #   - If `in_subdir` is a numeric value, the N first level directories will be removed.
-# - Patches named `sources/patches/${src_id}-*.patch` will be applied to `$dest_dir`
-# - Extra files in `sources/extra_files/$src_id` will be copied to dest_dir
+# - Patches named `patches/${src_id}-*.patch` will be applied to `$dest_dir`
```
Comment (Salamandar) [inline helpers/helpers.v2.1.d/utils:135]: Please use `patches/${src_id}/*.patch` instead, as by default git names patch files with a numbering prefix.

**Review by Salamandar — APPROVED** [review]

File: helpers/helpers.v1.d/mysql [hunk 3]
```diff
@@ -174,6 +174,19 @@ ynh_mysql_user_exists() {
...
+#
+# [internal]
+#
+# usage: ynh_mysql_database_exists database
+# | arg: database    - the database for which to check existence
+# | exit: Return 1 if the database doesn't exist, 0 otherwise
+#
+ynh_mysql_database_exists() {
+    local database=$1
+    mysqlshow | grep -q "^| $database "
```
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v1.d/mysql:187]:
```suggestion
    mysqlshow | grep -qE "^|\s+$database\s+|"
```

File: helpers/helpers.v2.1.d/php [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    [[ ! -e "$workdir/composer.phar" ]] || ynh_safe_rm $workdir/composer.phar
+
+    local composer_url="https://getcomposer.org/download/$composer_version/composer.phar"
+
+    # NB. we have to declare the var as local first,
+    # otherwise 'local foo=$(false) || echo 'pwet'" does'nt work
+    # because local always return 0 ...
+    local out
+    # Timeout option is here to enforce the timeout on dns query and tcp connect (c.f. man wget)
+    out=$(wget --tries 3 --no-dns-cache --timeout 900 --no-verbose --output-document=$workdir/composer.phar $composer_url 2>&1) \
```
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v2.1.d/php:323]: hmmmmaybe but not every app use the same composer version ... though i don't know what are the implications or if we should push for more standardization of composer versions (i barely know anything about it)

File: helpers/helpers.v2.1.d/utils [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+
+# Read the value of a key in the app's manifest
+#
+# usage: ynh_read_manifest "key"
+# | arg: key - Name of the key to find
+# | ret: the value associate to that key
+#
+# Requires YunoHost version 3.5.0 or higher.
+ynh_read_manifest() {
+    cat $YNH_APP_BASEDIR/manifest.toml | toml_to_json | jq ".$1" --raw-output
```
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v2.1.d/utils:459]: yeah, only in bookworm tho : https://packages.debian.org/bookworm/yq (or ... backports *spooky noises*)

File: helpers/helpers.v2.1.d/utils [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+# - Check the integrity with the specific sha256 sum
+# - Uncompress the archive to `$dest_dir`.
+#   - If `in_subdir` is true, the first level directory of the archive will be removed.
+#   - If `in_subdir` is a numeric value, the N first level directories will be removed.
+# - Patches named `patches/${src_id}-*.patch` will be applied to `$dest_dir`
+#
+# Requires YunoHost version 2.6.4 or higher.
+ynh_setup_source() {
+    # ============ Argument parsing =============
+    local -A args_array=([d]=dest_dir= [s]=source_id= [k]=keep= [r]=full_replace)
```
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v2.1.d/utils:140]: hmmmyeah agreed, just didn't want to go the rabbit hole of patching it but maybe we can still do it

File: helpers/helpers.v2.1.d/backup (hunk 1 above)
Comment (alexAubin) [reply helpers/helpers.v2.1.d/backup:30]: >Yeah that might be a bit restrictive and what if $data_dir/some_important_config_file.json ?

yeah idk, i tried to study carefully what was really done in the wild ... it's still unclear what policy we want exactly (for example it wasnt clear for log files when we discussed it on the chat the other day)

dunno what do about conf inside $data_dir ... but the main point is that the specific behavior basically happens only for safety-backups-before-upgrade (..or if people enabled backup_core_only manually...) hence the data_dir will be neither backuped nor removed (cf the --purge mechanism) -- and we REALLY have to get the two side consistent (the backup side and remove (prior to restore ofc). Currently there may be apps which do not declare their data_dir as "is_big" (sigh) hence might get simply rmed with no backup if an upgrade fails...

File: helpers/helpers.v2.1.d/utils [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+    [[ "$YNH_APP_UPGRADE_TYPE" == "UPGRADE_PACKAGE" ]] && return 1 || return 0
+}
+
+# Compare the current package version is strictly lower than another version given as an argument
+#
+# example: if ynh_app_upgrading_from_version_before 2.3.2~ynh1; then ...
+#
+# Requires YunoHost version 11.2 or higher.
+ynh_app_upgrading_from_version_before() {
+    local version=$1
```
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v2.1.d/utils:488]: meh idk i'm tired of having getopts to parse single args etc...

File: helpers/helpers.v2.1.d/php (hunk 4 above)
[the remark this answers is not in the document]
Comment (Salamandar) [reply helpers/helpers.v2.1.d/php:323]: Yeah see my PR, i just download composer to a file whose name contains the version. Also we can prevent multiple downloads.

File: helpers/helpers.v2.1.d/utils (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericgaspar) [reply helpers/helpers.v2.1.d/utils:140]: `full_replace` as default will break many apps

File: helpers/helpers.v2.1.d/utils (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexAubin) [reply helpers/helpers.v2.1.d/utils:488]: Yeah i don't know either, I was initially pretty enthusiast about "let's roll getopts with keyword args everywhere such that everything is uniform" but the more I looked at the code, the more it felt bloated with helper-calling-helpers-calling-helpers-calling-helpers, everytime re-parsing arguments and most of the time there's only a single really important and useful arg

I'm not that much concerned about performance, or possibly about keeping the logs more simple and less of a "huge unreadable soup"

But right now the more I look at the helpers and options, the more I could see a world where getopts would be the *exception* for the few cases where we really need several important options

I mean for example we have `ynh_print_info --message="foo bar"` and like, yeah, this is an helper to print an info, there's no other option, of course it's the message ...

`ynh_die` has `--message` and also `--ret_code` which in practice is never used so meh ...

Same kind of analysis goes for many helpers

**Pull request merged**